technologies/BehaviorTree.CPP/pyrobosim_btcpp/src/btcpp_executor.cpp

```diff
@@ -66,37 +70,76 @@ int main(int argc, char** argv)
   params.nh = nh;
   params.default_port_value = "execute_action";
 
+  factory.registerNodeType<BT::IsBatteryLow>("IsBatteryLow", nh->get_logger());
+  factory.registerNodeType<BT::IsBatteryFull>("IsBatteryFull", nh->get_logger());
   factory.registerNodeType<BT::CloseAction>("Close", params);
   factory.registerNodeType<BT::DetectObject>("DetectObject", params);
   factory.registerNodeType<BT::NavigateAction>("Navigate", params);
   factory.registerNodeType<BT::OpenAction>("Open", params);
   factory.registerNodeType<BT::PickObject>("PickObject", params);
   factory.registerNodeType<BT::PlaceObject>("PlaceObject", params);
 
-  // TO_WORKSHOP_USER: add here more rgistration, if you decided to implement your own nodes
+  // TO_WORKSHOP_USER: register here more Nodes, if you decided to implement your own
 
-  // optionally we can display and save the model of the tree
+  // Optionally, we can save the model of the Nodes registered in the factory
   if(save_model)
   {
     std::string xml_models = BT::writeTreeNodesModelXML(factory);
-    std::cout << "--------- Node Models ---------:\n" << xml_models << std::endl;
-    std::ofstream of("tree_nodes_model.xml");
+    std::ofstream of("pyrobosim_btcpp_model.xml");
     of << xml_models;
-    std::cout << "\nXML model of the tree saved in tree_nodes_model.xml\n" << std::endl;
+    RCLCPP_INFO(nh->get_logger(), "XML model of the tree saved in pyrobosim_btcpp_model.xml");
+    return 0;
   }
+  const std::filesystem::path filepath = GetFilePath(tree_filename);
 
   //----------------------------------
   // load a tree and execute
-  BT::Tree tree = factory.createTreeFromFile(filepath.string());
+  factory.registerBehaviorTreeFromFile(filepath.string());
+
+  // the global blackboard patterns is explaine here:
+  // https://www.behaviortree.dev/docs/tutorial-advanced/tutorial_16_global_blackboard
+  auto global_blackboard = BT::Blackboard::create();
+  BT::Tree tree = factory.createTree("MainTree", BT::Blackboard::create(global_blackboard));
 
   // This will add console messages for each action and condition executed
   BT::StdCoutLogger console_logger(tree);
   console_logger.enableTransitionToIdle(false);
 
-  // This is the "main loop":xecution is completed once the tick() method returns SUCCESS of FAILURE
-  BT::NodeStatus res = tree.tickWhileRunning();
+  BT::Groot2Publisher groot2_publisher(tree);
+
+  rclcpp::executors::SingleThreadedExecutor executor;
+  executor.add_node(nh);
+
+  bool state_received = false;
+
+  // create a subscriber to /robot/robot_state to update the blackboard
+  auto robot_state_subscriber = nh->create_subscription<pyrobosim_msgs::msg::RobotState>(
+      "/robot/robot_state", 10,
```

Comment:
So the `/robot` prefix is specific to there being a robot with that name in the simulator. There could just as well be a different name, or multiple robots. So can this be configurable?

Not needed for now, but more for your information.